File: backend_old/src/controllers/searchController.js
const db = require('../../config/db');

// GET /api/search?q=nombre&type=person&country=CL&limit=10
const search = async (req, res) => {
  const { q = '', type, country, limit = 10 } = req.query;

  if (q.trim().length < 2) {
    return res.status(400).json({ error: 'El término de búsqueda debe tener al menos 2 caracteres' });
  }

  try {
    // Construcción dinámica del WHERE
    const conditions = [`to_tsvector('spanish', name) @@ plainto_tsquery('spanish', $1)`];
    const params     = [q.trim()];
    let   idx        = 2;

    if (type) {
      conditions.push(`type = $${idx++}`);
      params.push(type);
    }
    if (country) {
      conditions.push(`country = $${idx++}`);
      params.push(country.toUpperCase());
    }

    params.push(parseInt(limit, 10));

    const sql = `
      SELECT
        id,
        external_id,
        type,
        name,
        country,
        risk_score,
        metadata
      FROM nodes
      WHERE ${conditions.join(' AND ')}
      ORDER BY risk_score DESC, name ASC
      LIMIT $${idx}
    `;

    const { rows } = await db.query(sql, params);
    res.json({ results: rows, total: rows.length });
  } catch (err) {
    console.error('Error en búsqueda:', err);
    res.status(500).json({ error: 'Error al buscar' });
  }
};

module.exports = { search };

File: backend/controllers/search_controller.py
from sqlalchemy import text
from sqlalchemy.orm import Session
from models.node import Node

async def search(
    db: Session,
    q: str,
    node_type: str = None,
    country: str = None,
    limit: int = 10
):
    """
    Búsqueda full-text en nodos de MySQL.
    Soporta búsqueda por nombre con LIKE.
    """
    if len(q) < 2:
        return {"error": "Query debe tener al menos 2 caracteres"}, 400

    query = db.query(Node)

    # Búsqueda por nombre (LIKE)
    query = query.filter(Node.name.ilike(f"%{q}%"))

    # Filtros opcionales
    if node_type:
        query = query.filter(Node.type == node_type)
    if country:
        query = query.filter(Node.country == country)

    # Orden por risk_score (descendente) y nombre (ascendente)
    results = query.order_by(
        Node.risk_score.desc(),
        Node.name.asc()
    ).limit(limit).all()

    return {
        "query": q,
        "results": [
            {
                "id": r.id,
                "external_id": r.external_id,
                "type": r.type,
                "name": r.name,
                "country": r.country,
                "metadata": r.meta or {},
                "risk_score": r.risk_score,
                "created_at": r.created_at.isoformat() if r.created_at else None
            }
            for r in results
        ],
        "total": len(results)
    }
